### PR TITLE
Altframe hooks

### PR DESCRIPTION
**Pull request by QuPsi**
Introduces ability to set the fence max altitude frame. Also includes @peterbarker's param-validate functionality.

### CODE REVIEW

**Comment by amilcarlucas** [comment]
Looks like you are reverting some changes from master by accident. My guess is that a couple of merge conflicts went south.
You should fix that first

File: libraries/AP_Common/Location.cpp [hunk 1]
```diff
@@ -107,39 +107,45 @@ bool Location::get_alt_cm(AltFrame desired_frame, int32_t &ret_alt_cm) const
...
     }
 #endif
-    Location::AltFrame frame = get_alt_frame();
+    Location::AltFrame current_frame = get_alt_frame();
 
+    return get_spec_alt_cm(current_frame, desired_frame, alt, ret_alt_cm);
+}
+
+// mday99: This function was a part of get_alt_cm up above. I pulled it out to accept any altitude as an input
+bool Location::get_spec_alt_cm(AltFrame current_frame, AltFrame desired_frame, int32_t alt_conv, int32_t &ret_alt_cm) const
```
Comment (peterbarker) [inline libraries/AP_Common/Location.cpp:116]:
```suggestion
bool Location::convert_altitude_frame(AltFrame current_frame, AltFrame desired_frame, int32_t alt_in_cm, int32_t &alt_out_cm) const
```

File: libraries/AP_Common/Location.h [hunk 2]
```diff
@@ -44,6 +44,8 @@ class Location
     // can only happen if the original frame or desired frame is above-terrain
     bool get_alt_cm(AltFrame desired_frame, int32_t &ret_alt_cm) const WARN_IF_UNUSED;
 
+    bool get_spec_alt_cm(AltFrame current_frame, AltFrame desired_frame, int32_t _alt_conv, int32_t &ret_alt_cm) const;
```
Comment (peterbarker) [inline libraries/AP_Common/Location.h:47]: This should be static, yes?

File: libraries/AP_Param/AP_Param.cpp [hunk 3]
```diff
@@ -36,6 +36,7 @@
 #if CONFIG_HAL_BOARD == HAL_BOARD_SITL
     #include <SITL/SITL.h>
 #endif
+#include <signal.h>
```
Comment (peterbarker) [inline libraries/AP_Param/AP_Param.cpp:39]: Why is this included?

File: ArduCopter/system.cpp [hunk 4]
```diff
@@ -13,6 +13,22 @@ static void failsafe_check_static()
     copter.failsafe_check();
 }
 
+static bool validate_fence_frame(float newframe) 
+{    
+    if ((int)newframe > 3 || (int)newframe < 0) {
```
Comment (peterbarker) [inline ArduCopter/system.cpp:18]: This would be better done as a `switch`, enumerating the valid values.

File: ArduCopter/system.cpp [hunk 5]
```diff
@@ -13,6 +13,22 @@ static void failsafe_check_static()
...
+        return false;
+    } else {
```
Comment (peterbarker) [inline ArduCopter/system.cpp:20]: Pointless `else`

File: ArduCopter/system.cpp [hunk 6]
```diff
@@ -13,6 +13,22 @@ static void failsafe_check_static()
...
+        return AP::fence()->conv_max_alt_frame((int)newframe);
```
Comment (peterbarker) [inline ArduCopter/system.cpp:21]: These validation routines were never meant to modify values - only validate them.

Coupling parmeters like this is going to lead to all sorts of confusion.  Pretty sure we want to leave it up to the user to set both the frame and altitude at the same time.

File: ArduPlane/Parameters.cpp [hunk 7]
```diff
@@ -967,41 +968,17 @@ const AP_Param::Info Plane::var_info[] = {
     // @Path: GCS_Mavlink.cpp
     GOBJECTN(_gcs.chan_parameters[0], gcs0,        "SR0_",     GCS_MAVLINK_Parameters),
 
-#if MAVLINK_COMM_NUM_BUFFERS >= 2
```
Comment (peterbarker) [inline ArduPlane/Parameters.cpp:970]: These are the changes you're reverting when you shouldn't be.

File: libraries/AC_Fence/AC_Fence.h [hunk 8]
```diff
@@ -43,6 +44,8 @@ class AC_Fence
 
     void init() {
         _poly_loader.init();
+        AP_Param::set_and_save_by_name("FENCE_ALT_MAX", AC_FENCE_ALT_MAX_DEFAULT);
+        AP_Param::set_and_save_by_name("FENCE_ALT_FRAME", AC_FENCE_ALT_FRAME_DEFAULT);
```
Comment (peterbarker) [inline libraries/AC_Fence/AC_Fence.h:48]: This definitely doesn't look right.  What happened to people's persistence here?!

File: libraries/AP_Param/AP_Param.h [hunk 9]
```diff
@@ -62,7 +64,7 @@
 
 // don't shift index 0 to index 63. Use this when you know there will be
 // no conflict with the parent
-#define AP_PARAM_FLAG_NO_SHIFT      (1<<3)
+#define AP_PARAM_FLAG_NO_SHIFT           (1<<3)
```
Comment (peterbarker) [inline libraries/AP_Param/AP_Param.h:67]: Bad whitespace change.

File: libraries/AP_Param/AP_Param.h [hunk 10]
```diff
@@ -24,6 +24,8 @@
 #include <stdint.h>
 #include <cmath>
 
+#include <stdio.h>
```
Comment (peterbarker) [inline libraries/AP_Param/AP_Param.h:27]: Why?

File: libraries/GCS_MAVLink/GCS_Param.cpp [hunk 11]
```diff
@@ -71,7 +71,7 @@ GCS_MAVLINK::queued_param_send()
     }
     count -= async_replies_sent_count;
 
-    while (count && _queued_parameter != nullptr && get_last_txbuf() > 50) {
```
Comment (peterbarker) [inline libraries/GCS_MAVLink/GCS_Param.cpp:74]: Another master change reverted.  Not good.

**Review by peterbarker — CHANGES_REQUESTED** [review]
I can see some uses, but I am curious - what is your need to be able to specify the altitude frame of the altitude fences?

**Comment by amilcarlucas** [comment]
Do not merge master into this branch. Instead rebase this branch on top of master.
Instructions are here: https://ardupilot.org/dev/docs/git-interactive-rebase.html

**Comment by peterbarker** [comment]
Closing in favour of https://github.com/ArduPilot/ardupilot/pull/18038

**Pull request closed without merging**